californium-core/src/main/java/org/eclipse/californium/core/network/serialization/TcpDataSerializer.java

```diff
@@ -0,0 +1,166 @@
+/*******************************************************************************
+ * Copyright (c) 2015, 2016 Institute for Pervasive Computing, ETH Zurich and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Matthias Kovatsch - creator and main architect
+ *    Martin Lanter - architect and re-implementation
+ *    Dominique Im Obersteg - parsers and initial implementation
+ *    Daniel Pauli - parsers and initial implementation
+ *    Kai Hudalla - logging
+ *    Bosch Software Innovations GmbH - turn into utility class with static methods only
+ *    Joe Magerramov (Amazon Web Services) - CoAP over TCP support.
+ ******************************************************************************/
+package org.eclipse.californium.core.network.serialization;
+
+import org.eclipse.californium.core.coap.CoAP.Code;
+import org.eclipse.californium.core.coap.Message;
+import org.eclipse.californium.core.coap.Option;
+import org.eclipse.californium.core.coap.Request;
+import org.eclipse.californium.core.coap.Response;
+
+import java.util.List;
+
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.CODE_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.LENGTH_NIBBLE_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.MESSAGE_ID_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.OPTION_DELTA_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.OPTION_LENGTH_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.PAYLOAD_MARKER;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.TOKEN_LENGTH_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.TYPE_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.VERSION;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.VERSION_BITS;
+
+/**
+ * The DataSerialized serializes outgoing messages to byte arrays based on CoAP TCP/TLS spec:
+ * <a href="https://tools.ietf.org/html/draft-ietf-core-coap-tcp-tls"/>
+ */
+public final class TcpDataSerializer implements DataSerializer {
+
+	@Override
+	public byte[] serializeRequest(Request request) {
+		DatagramWriter writer = new DatagramWriter();
+		Code code = request.getCode();
+		serializeMessage(writer, request, code == null ? 0 : code.value);
```

Comment:
I know that Request allows for instantiation with null Code but I have yet to see why that would be necessary ... do you have any idea?